language-support/scala/bindings/src/main/scala/com/digitalasset/ledger/client/binding/binding/EnumCompanion.scala

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.ledger.client.binding
+
+import com.digitalasset.ledger.api.v1.{value => rpcvalue}
+import com.digitalasset.ledger.client.binding.encoding.{LfEncodable, LfTypeEncoding}
+import scalaz.Liskov.<~<
+import scalaz.OneAnd
+import scalaz.syntax.functor._
+import scalaz.std.vector._
+
+abstract class EnumCompanion[T](implicit isEnum: T <~< EnumRef) extends ValueRefCompanion {
+
+  val firstValue: T
+  val otherValues: Vector[T]
+
+  final def values: Vector[T] = firstValue +: otherValues
+
+  implicit final lazy val `the enum Value`: Value[T] = new `Value ValueRef`[T] {
+    private[this] val readers = values.map(e => (isEnum(e).constructor: String) -> e).toMap
+
+    override def read(argValue: rpcvalue.Value.Sum): Option[T] =
+      argValue.enum flatMap (e => readers.get(e.constructor))
+
+    private[this] val rpcValues = values.map(e => ` enum`(isEnum(e).constructor))
+
+    override def write(enum: T): rpcvalue.Value.Sum =
+      rpcValues(isEnum(enum).index)
+  }
+
+  @SuppressWarnings(Array("org.wartremover.warts.Any"))
+  implicit final val `the enum LfEncodable`: LfEncodable[T] = new LfEncodable[T] {
+    override def encoding(lte: LfTypeEncoding): lte.Out[T] =
+      lte.enumAll(
+        ` dataTypeId`,
+        (firstValue +: otherValues).map(isEnum(_).constructor).zipWithIndex.toMap,
+        OneAnd(firstValue, otherValues).map(x => isEnum(x).constructor -> x))
```

Comment:
Doesn't refer to `lte`, so can be pulled out like `readers` and `rpcValues` (with the same `lazy val` for the containing `val`).

Reply:
Done.